server/src/main/java/org/elasticsearch/repositories/ShardSnapshotContext.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories;
+
+import org.apache.lucene.index.IndexCommit;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.snapshots.IndexShardSnapshotStatus;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.snapshots.SnapshotId;
+
+import java.util.Map;
+
+/**
+ * Context holding the state for creating a shard snapshot via {@link Repository#snapshotShard(ShardSnapshotContext)}.
+ * Wraps a {@link org.elasticsearch.index.engine.Engine.IndexCommitRef} that is released once this instances is completed by invoking
+ * either its {@link #onResponse(ShardSnapshotResult)} or {@link #onFailure(Exception)} callback.
+ */
+public final class ShardSnapshotContext extends ActionListener.Delegating<ShardSnapshotResult, ShardSnapshotResult> {
```

Comment:
Style nit: could we keep the `listener` out of this? That way the `snapshotShard` method remains obviously async (it takes a listener). Otherwise the chances are high that we'll forget to complete this context in some future refactoring.

Also `snapshotShard(ShardSnapshotContext ...)` could we have the words `shard` and `snapshot` the same way round?

Reply:
I did the renaming, but I'd rather not take the `listener` out of things. If we move the listener out of this object, we don't get to dry up the resource handling. Also, the listener/state handling in this method is already somewhat complicated because we not only have the listener to worry about, we also have to go through the correct state steps on the `org.elasticsearch.index.snapshots.IndexShardSnapshotStatus` so it seems less an implementation would forget the completing the `context` (especially when it still implements `ActionListener`)  when it gets all the other steps right?

Reply:
ok, thanks for considering it anyway :)